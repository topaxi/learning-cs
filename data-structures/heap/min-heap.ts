import { Heap } from './heap'

export class MinHeap<T> extends Heap<T> {}
