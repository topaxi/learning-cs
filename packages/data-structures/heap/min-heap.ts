import { Heap } from './heap'

export class MinHeap<T = number> extends Heap<T> {
  constructor(comparator = (a: T, b: T): number => (a as any) - (b as any)) {
    super(comparator)
  }
}
