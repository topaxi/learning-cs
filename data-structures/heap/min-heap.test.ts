import { MinHeap } from './min-heap'

describe('MinHeap', () => {
  test('should create empty heap', () => {
    let heap = new MinHeap()

    expect(heap.peek()).toBeNull()
    expect(heap.empty).toBe(true)
  })

  test('should push value to heap', () => {
    let heap = new MinHeap()

    heap.push(42)

    expect(heap.peek()).toBe(42)
    expect(heap.empty).toBe(false)
  })

  test('should push values to heap', () => {
    let heap = new MinHeap()

    heap.push(43)
    heap.push(44)
    heap.push(46)
    heap.push(42)
    heap.push(44)

    expect(heap).toMatchInlineSnapshot(`
            Array [
              42,
              43,
              46,
              44,
              44,
            ]
        `)

    expect(Array.from(heap.consume())).toMatchInlineSnapshot(`
                  Array [
                    42,
                    43,
                    44,
                    44,
                    46,
                  ]
            `)
  })

  test('should pop values from the heap', () => {
    let heap = new MinHeap()

    heap.push(43)
    heap.push(44)
    heap.push(42)
    heap.push(46)
    heap.push(42)

    expect(heap).toMatchInlineSnapshot(`
      Array [
        42,
        42,
        43,
        46,
        44,
      ]
    `)

    expect(heap.pop()).toBe(42)

    expect(heap).toMatchInlineSnapshot(`
      Array [
        42,
        44,
        43,
        46,
      ]
    `)

    expect(heap.pop()).toBe(42)

    expect(heap).toMatchInlineSnapshot(`
            Array [
              43,
              44,
              46,
            ]
        `)

    expect(Array.from(heap.consume())).toMatchInlineSnapshot(`
                        Array [
                          43,
                          44,
                          46,
                        ]
                `)
  })
})
