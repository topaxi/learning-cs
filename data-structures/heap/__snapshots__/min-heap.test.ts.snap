// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MinHeap should push values to heap 1`] = `
Array [
  42,
  43,
  46,
  44,
]
`;
